app/controllers/exercises_controller.rb

```diff
@@ -122,60 +121,97 @@ def requests_for_comments
+          # exercise_found: codeharbor_check[:exercise_found],
+          update_right: codeharbor_check[:update_right],
+          error: codeharbor_check[:error],
+          exported: false
+        }
+      )
+    }, status: 200
+  end
+
+  def export_external_confirm
+    push_type = params[:push_type]
+
+    return render json: {}, status: 500 unless %w[create_new export].include? push_type
+
+    @exercise.uuid = SecureRandom.uuid if push_type == 'create_new'
+
+    error = ExerciseService::PushExternal.call(
+      zip: ProformaService::ExportTask.call(exercise: @exercise),
+      codeharbor_link: current_user.codeharbor_link
+    )
+    if error.nil?
+      render json: {
+        status: 'success',
+        message: t('exercises.export_codeharbor.successfully_exported', id: @exercise.id, title: @exercise.title),
+        actions: render_to_string(partial: 'export_actions', locals: {exercise: @exercise, exported: true, error: error})
+      }
+    else
+      render json: {
+        status: 'fail',
+        message: t('exercises.export_codeharbor.export_failed', id: @exercise.id, title: @exercise.title, error: error),
+        actions: render_to_string(partial: 'export_actions', locals: {exercise: @exercise, exported: true, error: error})
+      }
     end
   end
 
-  def user_for_oauth2_request
-    authorizationHeader = request.headers['Authorization']
-    if authorizationHeader == nil
-      raise ({status: 401, message: 'No Authorization header'})
-    end
+  def import_uuid_check
+    user = user_from_api_key
+    return render json: {}, status: 401 if user.nil?
 
-    oauth2Token = authorizationHeader.split(' ')[1]
-    if oauth2Token == nil || oauth2Token.size == 0
-      raise ({status: 401, message: 'No token in Authorization header'})
-    end
+    uuid = params[:uuid]
+    exercise = Exercise.find_by(uuid: uuid)
 
-    user = user_by_code_harbor_token(oauth2Token)
-    if user == nil
-      raise ({status: 401, message: 'Unknown OAuth2 token'})
-    end
+    return render json: {exercise_found: false, message: t('exercises.import_codeharbor.check.no_exercise')} if exercise.nil?
+    return render json: {exercise_found: true, update_right: false, message: t('exercises.import_codeharbor.check.exercise_found_no_right')} unless ExercisePolicy.new(user, exercise).update?
 
-    return user
+    render json: {exercise_found: true, update_right: true, message: t('exercises.import_codeharbor.check.exercise_found')}
   end
-  private :user_for_oauth2_request
 
-  def user_by_code_harbor_token(oauth2Token)
-    link = CodeHarborLink.where(:oauth2token => oauth2Token)[0]
-    if link != nil
-      return link.user
+  def import_exercise
+    tempfile = Tempfile.new('codeharbor_import.zip')
+    tempfile.write request.body.read.force_encoding('UTF-8')
+    tempfile.rewind
+
+    user = user_from_api_key
+    return render json: {}, status: 401 if user.nil?
+
+    exercise = nil
+    ActiveRecord::Base.transaction do
```

Comment:
During import the `ConvertTaskToExercise` service changes a lot of attributes not only on the exercise object but also its dependents (files).
This could lead to a scenario where some (or all) files are deleted in the service and the exercise is not valid anymore. The transaction can restore the files in that case.

I also managed to add a spec to reflect that edge-case

Reply:
👍